### PR TITLE
[pt1][quant] Add print function for QTensor

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19513 [pt1][quant] Add print function for QTensor**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15017168/)

Add support for printing a QTensor in python frontend

Differential Revision: [D15017168](https://our.internmc.facebook.com/intern/diff/D15017168/)

### CODE REVIEW

File: torch/_tensor_str.py [hunk 1]
```diff
@@ -285,7 +302,7 @@ def _str(self):
             else:
                 tensor_str = _tensor_str(self, indent)
 
-    if self.layout != torch.strided:
+    if not self.is_quantized and self.layout != torch.strided:
```
Comment (gchanan) [inline torch/_tensor_str.py]: why do you need this? is quantized a layout?

File: test/test_torch.py [hunk 2]
```diff
@@ -2683,6 +2683,11 @@ def test_qtensor(self):
         # This calls _th_fill_
         # qr[0] = 8 # float asignment
         # self.assertEqual(qr.item(), 8)
+        # we can also print a qtensor
+        print(qr)
```
Comment (gchanan) [inline test/test_torch.py]: can you just assert what you expect the print to look like?

File: torch/_tensor_str.py [hunk 3]
```diff
@@ -266,6 +266,23 @@ def _str(self):
         if values.numel() == 0:
             values_str += ', size=' + str(tuple(values.shape))
         tensor_str = indices_prefix + indices_str + '),\n' + ' ' * indent + values_prefix + values_str + ')'
+    elif self.is_quantized:
+        suffixes.append('size=' + str(tuple(self.shape)))
+        if not has_default_dtype:
```
Comment (gchanan) [inline torch/_tensor_str.py]: can you qtensor ever be the default dtype?  (It shouldn't be, but I don't know if it's checked).

File: torch/_tensor_str.py [hunk 4]
```diff
@@ -266,6 +266,23 @@ def _str(self):
...
+        values_prefix = 'float_values=tensor('
+        values = self.dequantize()
+        values_str = _tensor_str(values, indent + len(values_prefix))
+        if values.numel() == 0:
+            values_str += ', size=' + str(tuple(values.shape))
+        int_repr_prefix = 'int_repr=tensor('
+        int_repr = self.int_repr()
+        int_repr_str = _tensor_str(int_repr, indent + len(int_repr_prefix))
+        if int_repr.numel() == 0:
+            int_repr_str += ', size=' + str(tuple(int_repr.shape))
```
Comment (gchanan) [inline torch/_tensor_str.py]: why can't you handle the size things automatically through the recursive calls?

**Review by gchanan — CHANGES_REQUESTED** [review]
why are we printing both the float values and the int values always?  I'm not against it, but would want to hear the reasoning.

File: test/test_torch.py (hunk 2 above)
Comment (jerryzh168) [reply test/test_torch.py]: sure

File: torch/_tensor_str.py (hunk 3 above)
Comment (jerryzh168) [reply torch/_tensor_str.py]: probably not

File: torch/_tensor_str.py (hunk 4 above)
Comment (jerryzh168) [reply torch/_tensor_str.py]: oh this is copied from sparse, we don't need this for quantized.

File: torch/_tensor_str.py (hunk 1 above)
Comment (jerryzh168) [reply torch/_tensor_str.py]: layout was not registered, I'll add it.

**Comment by jerryzh168** [comment]
@gchanan can you review again

File: aten/src/ATen/native/quantized/QTensor.cpp [hunk 5]
```diff
@@ -36,7 +36,9 @@ Tensor int_repr_quant(const Tensor& self) {
   Tensor dst = at::empty(self.sizes(), self.options().dtype(at::kByte));
   uint8_t* self_data = reinterpret_cast<uint8_t *>(self.data<qint8>());
   uint8_t* dst_data = dst.data<uint8_t>();
-  memcpy(dst_data, self_data, self.numel());
+  if (self.numel() > 0) {
+    memcpy(dst_data, self_data, self.numel());
```
Comment (gchanan) [inline aten/src/ATen/native/quantized/QTensor.cpp]: this is due to asan issues?

File: test/test_torch.py [hunk 6]
```diff
@@ -2683,6 +2683,17 @@ def test_qtensor(self):
         # This calls _th_fill_
         # qr[0] = 8 # float asignment
         # self.assertEqual(qr.item(), 8)
+        # we can also print a qtensor
+        self.assertEqual(str(qr),
+                         "tensor(is_quantized=True, qscheme=per_tensor_affine,\n" +
```
Comment (gchanan) [inline test/test_torch.py]: did you respond to the question about why we are printing both?

Comment (gchanan) [inline test/test_torch.py]: why do you need to print is_quantized instead of just the dtype?  The dtype implies it's quantized, correct?

Comment (gchanan) [reply test/test_torch.py]: I'd also move all of the metadata (e.g. scale, zero_point), to the end of the print, like it is currently with non-quantized tensors.

**Review by gchanan — CHANGES_REQUESTED** [review]

**Comment by jerryzh168** [comment]
@gchanan for printing both, I don't have a strong reason for it, mostly just assuming user might want to see both of them since they both represents the values that the QTensor contains.

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 5 above)
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/QTensor.cpp]: when numel() is 0, we are passing nullptr to memcpy

File: test/test_torch.py (hunk 6 above)
Comment (jerryzh168) [reply test/test_torch.py]: see comment

Comment (jerryzh168) [reply test/test_torch.py]: The way we check for whether a QTensor is quantized internally is checking whether the underlying TensorTypeId is QuantizedCPUTensorId rather than whether the data type is QInt, so I feel it might be better to print is_quantized.

Comment (jerryzh168) [reply test/test_torch.py]: probably we don't need that, qscheme already tells it's a QTensor

**Comment by jerryzh168** [comment]
@gchanan could you review again?

File: aten/src/ATen/native/quantized/QTensor.cpp (hunk 5 above)
Comment (gchanan) [reply aten/src/ATen/native/quantized/QTensor.cpp]: sorry, I meant ubsan.

File: torch/_tensor_str.py [hunk 7]
```diff
@@ -266,6 +266,20 @@ def _str(self):
         if values.numel() == 0:
             values_str += ', size=' + str(tuple(values.shape))
         tensor_str = indices_prefix + indices_str + '),\n' + ' ' * indent + values_prefix + values_str + ')'
+    elif self.is_quantized:
+        suffixes.append('size=' + str(tuple(self.shape)))
+        suffixes.append('dtype=' + str(self.dtype))
```
Comment (gchanan) [inline torch/_tensor_str.py]: nit: this should have a `if not has_default_dtype:` even though that can't currently be false.

File: torch/_tensor_str.py [hunk 8]
```diff
@@ -266,6 +266,20 @@ def _str(self):
...
+        suffixes.append('quantization_scheme=' + 'per_tensor_affine')
```
Comment (gchanan) [inline torch/_tensor_str.py]: there's no way other place to get the quantization scheme currently?  Like as an enum or something (where we could have another function that maps them to strings)?  Putting this in the print function doesn't seem good.

File: torch/_tensor_str.py [hunk 9]
```diff
@@ -266,6 +266,20 @@ def _str(self):
...
+        suffixes.append('scale=' + str(self.q_scale()))
```
Comment (gchanan) [inline torch/_tensor_str.py]: wait, why are these tensors at the python level?

File: torch/_tensor_str.py [hunk 10]
```diff
@@ -266,6 +266,20 @@ def _str(self):
...
+        suffixes.append('zero_point=' + str(self.q_zero_point()))
+        values_prefix = 'float_values=tensor('
```
Comment (gchanan) [inline torch/_tensor_str.py]: I don't really find the "print both" rationale convincing (outside of debugging, but there you can print whatever you want).

The way I think about this is as follows:
Ideally, the print should be as close to usable as possible (if you print a small strided tensor you can actually copy-paste the print and get the tensor out).  This doesn't work in all conditions (i.e. very large tensors), but it's a good guiding principle.

Now, the quantized APIs are still subject to change, but I think it's reasonable to say no one would ever pass both the float and integer values together.  Since the usual use case is quantizing a float tensor, I'd lean towards just printing the float values (and you don't have to label them `float_values`).

Now, maybe this is a little strange when you combine with the zero_point and scale, because those are already "included" in the float values, but I'd really view those as should be belonging to the dtype, which we always print in this case.

So in summary, if you don't feel strongly otherwise, I would suggest:
- not printing int_repr.
- not including float_values
- not printing scale/zero point as tensors, but just as values.

File: torch/_tensor_str.py (hunk 8 above)
Comment (jerryzh168) [reply torch/_tensor_str.py]: Yeah, we'll change that when we add q_scheme(), I'll add a comment

File: torch/_tensor_str.py (hunk 9 above)
Comment (jerryzh168) [reply torch/_tensor_str.py]: I don't know..It is Scalar in API, but do we have Variable -> Scalar? I suppose self.q_scale() is a variable?

Comment (gchanan) [reply torch/_tensor_str.py]: I think you can just do `self.q_scale().item()`.

File: torch/_tensor_str.py (hunk 10 above)
Comment (jerryzh168) [reply torch/_tensor_str.py]: Sure

File: torch/_tensor_str.py [hunk 11]
```diff
@@ -266,6 +266,15 @@ def _str(self):
         if values.numel() == 0:
             values_str += ', size=' + str(tuple(values.shape))
         tensor_str = indices_prefix + indices_str + '),\n' + ' ' * indent + values_prefix + values_str + ')'
+    elif self.is_quantized:
+        suffixes.append('size=' + str(tuple(self.shape)))
+        if not has_default_dtype:
+            suffixes.append('dtype=' + str(self.dtype))
+        # TODO: change to a call to self.q_scheme() when we add q_scheme method
+        suffixes.append('quantization_scheme=' + 'per_tensor_affine')
```
Comment (gchanan) [inline torch/_tensor_str.py]: can we just do a minimal q_scheme thing in this PR?  The issue with this is there is no enforcement mechanism to make us realize something is wrong once we support different qschemes.

Or heck, just make it not print the quantization scheme for now.

Comment (jerryzh168) [reply torch/_tensor_str.py]: Sure, let me comment this for now

**Review by gchanan — APPROVED** [review]
lgtm.
